caffe2/proto/caffe2.proto

```diff
@@ -219,6 +270,11 @@ message NetDef {
   // blobs' contents may be overwritten.
   repeated string external_input = 7;
   repeated string external_output = 8;
+
+  // List of profiles that modify operator arguments based on characteristics
+  // of the target system. Profiles are applied in the order in which they
+  // are listed.
+  repeated Profile profile = 9;
```

Comment:
I personally feel this is a hack. Protobufs are meant to be very rarely updated. The Predicate concept is very vaque. We will have to keep adding stuff there if we decide to go with protobuf changing approach. So I would discourage emulating predicates on the protobuf level. 

I wonder what are technical difficulties you see when you implement this without any changes to the protobuf. So far engines + runtime dispatch based on the available hardware worked. You just need to write dispatch logic on the library code level, but not on the serialization level. For example, check out perfkernels code, this is dispatch Dima and Yangqing added last year to support fast embedding lookup kernels from Misha.